File: dcc-portal-server/src/main/java/org/icgc/dcc/portal/server/util/Filters.java
/*
 * Copyright (c) 2014 The Ontario Institute for Cancer Research. All rights reserved.                             
 *                                                                                                               
 * This program and the accompanying materials are made available under the terms of the GNU Public License v3.0.
 * You should have received a copy of the GNU General Public License along with                                  
 * this program. If not, see <http://www.gnu.org/licenses/>.                                                     
 *                                                                                                               
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY                           
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES                          
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT                           
 * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,                                
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED                          
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;                               
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER                              
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN                         
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.icgc.dcc.portal.server.util;

import static com.google.common.base.Preconditions.checkState;
import static lombok.AccessLevel.PRIVATE;
import static org.icgc.dcc.portal.server.model.IndexModel.ALL;
import static org.icgc.dcc.portal.server.model.IndexModel.IS;
import static org.icgc.dcc.portal.server.model.EntityType.GENE;
import static org.icgc.dcc.portal.server.pql.convert.FiltersConverter.ENTITY_SET_PREFIX;
import static org.icgc.dcc.portal.server.util.JsonUtils.MAPPER;

import java.util.List;
import java.util.UUID;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.POJONode;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import com.google.common.collect.Sets;

import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.val;

/**
 * API filter utilities, not to be confused with Elasticsearch filters.
 */
@NoArgsConstructor(access = PRIVATE)
public final class Filters {

  private static final String ID = "id";

  public static ObjectNode emptyFilter() {
    return MAPPER.createObjectNode();
  }

  public static ObjectNode andFilter(@NonNull List<ObjectNode> filters) {
    return andFilter(filters.toArray(new ObjectNode[filters.size()]));
  }

  public static ObjectNode andFilter(@NonNull ObjectNode... filters) {
    JsonNode left = filters[0];
    for (int i = 1; i < filters.length; i++) {
      val right = filters[i];

      left = andFilter(left, right);
    }

    return (ObjectNode) left;
  }

  public static ObjectNode entityFilter(@NonNull String entityName) {
    val entityFilter = emptyFilter();
    entityFilter.with(entityName);

    return entityFilter;
  }

  public static ObjectNode geneFilter() {
    return entityFilter(GENE.getId());
  }

  public static ObjectNode geneSetFilter(@NonNull String geneSetId) {
    val geneFilter = geneFilter();
    geneFilter.with("gene").set("geneSetId", is(geneSetId));

    return geneFilter;
  }

  public static ObjectNode geneTypeFilter(@NonNull String type) {
    val geneFilter = geneFilter();
    geneFilter.with("gene").set("type", is(type));

    return geneFilter;
  }

  public static ObjectNode goTermFilter(@NonNull String goTermId) {
    val geneFilter = geneFilter();
    geneFilter.with(GENE.getId()).set("goTermId", is(goTermId));

    return geneFilter;
  }

  public static ObjectNode pathwayFilter() {
    val geneFilter = geneFilter();
    geneFilter.with("gene").put("hasPathway", true);

    return geneFilter;
  }

  public static ObjectNode inputGeneSetIdFilter(@NonNull String inputGeneSetId) {
    val geneFilter = geneFilter();
    geneFilter.with("gene").set(ID, is(ENTITY_SET_PREFIX + inputGeneSetId));

    return geneFilter;
  }

  public static ObjectNode inputGeneSetFilter(@NonNull UUID inputGeneSetId) {
    // Method appears to be used by multiple contexts, with and without `ES:`
    val setId = inputGeneSetId.toString();
    val id = setId.startsWith(ENTITY_SET_PREFIX) ? setId : ENTITY_SET_PREFIX + setId;

    val inputGeneSetFilter = geneFilter();
    inputGeneSetFilter.with(GENE.getId()).set(ID, is(id));

    return inputGeneSetFilter;
  }

  public static ObjectNode is(@NonNull String value) {
    val is = emptyFilter();
    is.withArray(IS).add(value);

    return is;
  }

  public static ObjectNode mergeAnalysisUniverse(@NonNull ObjectNode current, @NonNull ObjectNode universe) {
    val result = current.deepCopy();
    if (universe.path("gene").path("goTermId").isMissingNode() == false) {
      val universeId = universe.get("gene").get("goTermId").withArray("is").get(0).asText();

      if (result.path("gene").path("goTermId").path("is").isMissingNode() == false) {

        for (val t : result.with("gene").with("goTermId").withArray("is")) {
          result.with("gene").with("goTermId").withArray(ALL).add(t);
        }
        result.with("gene").with("goTermId").withArray(ALL).add(universeId);
        result.with("gene").with("goTermId").remove(IS);

      } else {
        result.with("gene").with("goTermId").withArray(IS).add(universeId);
      }
    } else if (universe.path("gene").path("hasPathway").isMissingNode() == false) {
      result.with("gene").put("hasPathway", true);
    }
    // } else if (universe.path("gene").path(Universe.REACTOME_PATHWAYS.getGeneSetFacetName()).isMissingNode() == false)
    // {
    // result.with("gene").put(Universe.REACTOME_PATHWAYS.getGeneSetFacetName(), true);
    // }
    return result;
  }

  public static ObjectNode mergeAnalysisInputGeneList(@NonNull ObjectNode current, @NonNull ObjectNode geneEntitySet) {
    val result = current.deepCopy();
    if (geneEntitySet.path("gene").path("id").isMissingNode() == false) {
      val entitySetId = geneEntitySet.path("gene").path(ID).withArray(IS).get(0).asText();
      result.with("gene").set("id", is(entitySetId));
    }
    return result;
  }

  public static ObjectNode mergeAnalysisGeneSetFilter(@NonNull ObjectNode current, @NonNull ObjectNode geneSet) {
    val result = current.deepCopy();
    if (geneSet.path("gene").path("geneSetId").isMissingNode() == false) {
      val geneSetId = geneSet.get("gene").get("geneSetId").withArray("is").get(0).asText();

      if (result.path("gene").path("geneSetId").path("is").isMissingNode() == false) {
        for (val t : result.with("gene").with("geneSetId").withArray("is")) {
          result.with("gene").with("geneSetId").withArray(ALL).add(t);
        }
        result.with("gene").with("geneSetId").withArray(ALL).add(geneSetId);
        result.with("gene").with("geneSetId").remove("is");
      } else {
        result.with("gene").with("geneSetId").withArray(IS).add(geneSetId);
      }
    }
    return result;
  }

  private static JsonNode andFilter(JsonNode left, JsonNode right) {
    val and = emptyFilter();

    if (right.getNodeType() == left.getNodeType()) {
      // Same types

      // Arbitrary
      val field = left;

      if (field.isObject()) {
        val fieldNames = ImmutableSet.<String> builder().addAll(left.fieldNames()).addAll(right.fieldNames()).build();
        for (val fieldName : fieldNames) {
          val leftField = left.path(fieldName);
          val rightField = right.path(fieldName);
          val andField = andFilter(leftField, rightField);

          and.set(fieldName, andField);
        }
      } else if (field.isArray()) {
        val values = Sets.<Object> newLinkedHashSet();
        for (val element : Iterables.concat(left, right)) {
          if (element.isNumber()) {
            values.add(element.asInt());
          } else if (element.isBoolean()) {
            values.add(element.asBoolean());
          } else if (element.isTextual()) {
            values.add(element.asText());
          } else if (element.isPojo()) {
            val pojo = (POJONode) element;
            val value = pojo.getPojo();
            values.add(value);
          } else {
            values.add(element);
          }
        }

        val result = MAPPER.createArrayNode();
        for (val value : values) {
          result.addPOJO(value);
        }

        return result;
      } else if (field.isValueNode()) {
        val result = MAPPER.createArrayNode();
        result.add(left);
        result.add(right);

        return result;
      } else if (field.isMissingNode()) {
        // Can't happen - TODO: Explain Why it cannot happen. Also empty branch???
      }
    } else {
      // Different types
      if (left.isMissingNode()) {
        return right;
      } else if (right.isMissingNode()) {
        return left;
      } else {
        checkState(false, "Node type(s) not expected: %s, %s", left, right);
      }
    }

    return and;
  }

}

File: dcc-portal-ui/app/scripts/common/js/filters.js
/*
 * Copyright 2016(c) The Ontario Institute for Cancer Research. All rights reserved.
 *
 * This program and the accompanying materials are made available under the terms of the GNU Public
 * License v3.0. You should have received a copy of the GNU General Public License along with this
 * program. If not, see <http://www.gnu.org/licenses/>.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

'use strict';

angular.module('icgc.common.filters', [])
  .constant('filterConstants', {
    FILTER_NAME: 'filters',
    FILTER_EVENTS: {
      FILTER_UPDATE_EVENT: 'event.filter.updated'
    }
  })
  .service('FilterService', function($location, $rootScope, filterConstants) {

    var _service = this,
        _filtersObj = {};

    _service.filters = _filters;
    _service.removeFilters = _removeFilters;
    _service.getCachedFiltersFactory = _getCachedFiltersFactory;
    _service.filterParam = _filterParam;
    _service.mergeIntoFilters = _mergeIntoFilters;
    _service.overwriteFiltersAtObjectLevel = _overwriteFiltersAtObjectLevel;
    _service.merge = _merge;
    _service.constants = filterConstants;


    ////////////////////////////////////////////////////////////////

    function _getLocationFiltersJSON() {
      var filters = _.get($location.search(), filterConstants.FILTER_NAME, '{}');
      return filters;
    }

    function _setFilterJSONAndNotify(filtersJSON) {
      _filtersObj = JSON.parse(filtersJSON);
      _notifyFilterChangeEvent();
    }

    /**
     * Initializes the filter service
     * @private
     */
    function _init() {
      $rootScope.$watch(_getLocationFiltersJSON, _setFilterJSONAndNotify);

      // On state changes ensure that we have the correct filters
      $rootScope.$on('$stateChangeSuccess', function() {

          _filtersObj = JSON.parse(_getLocationFiltersJSON());

      });
    }

    /**
     *
     * @param filters
     * @returns {copy of _filtersObj}
     * @private
     */
    function _filters(filters) {

      if (arguments.length === 1) {

        if (_.isObject(filters)) {
          _filtersObj = _.cloneDeep(filters);
        }
        else {
          _filtersObj = {};
        }
        _updateFilterParamsURL();
      }

      return _.cloneDeep(_filtersObj); // Do not allow overwriting the original object!
    }

    /**
     * Removes all filters
     * @returns {*}
     * @private
     */
    function _removeFilters() {
      return _filters(null);
    }

    /**
     * Setter/Getter for a single filter parameter
     * @param filterKey
     * @param filterVal
     * @returns {_filtersObj[filterKey]|null}
     * @private
     */
    function _filterParam(filterKey, filterVal) {

      if (arguments.length === 2) {
        _filtersObj[filterKey] = filterVal;
        _updateFilterParamsURL();
      }

      return _filtersObj[filterKey] || null;
    }

    /**
     * Updates the url on filter based on the current state of the filter object
     * @private
     */
    function _updateFilterParamsURL() {
      var filterVal = null;

      if (! _.isEmpty(_filtersObj)) {
        filterVal = JSON.stringify(_filtersObj);
      }

      $location.search(filterConstants.FILTER_NAME, filterVal);
    }

    /**
     * Return a factory object which can be used to return a cached version of the filters - useful for long running
     * http requests that require multiple requests over a period of time.
     * @returns {{_cachedFilters: {}, filters: filterService.filters, updateCache: filterService.updateCache}}
     * @private
     */
    function _getCachedFiltersFactory() {
      var initialFilters = _filters();

      return {
          _cachedFilters: initialFilters,
          filters: function(filterObj) {

            if (arguments.length === 0) {
              return _.cloneDeep(this._cachedFilters);
            }
            else {
              this._cachedFilters = filterObj;
            }

            return this._cachedFilters;
        },
        updateCache: function(newFilters) {
          var filterObj = newFilters && typeof newFilters === 'object' ? newFilters : _filters();
          this.filters(filterObj);
        }
      };
    }

    /**
     * Notifies observers of a filter change event.
     * @private
     */
    function _notifyFilterChangeEvent() {
      var filterNotifyObj = {
        currentFilters: _.cloneDeep(_filtersObj),
        currentSearchParams: $location.search(),
        currentPath: $location.path()
      };

      $rootScope.$broadcast(filterConstants.FILTER_EVENTS.FILTER_UPDATE_EVENT, filterNotifyObj);
    }

    function _overwriteFiltersAtObjectLevel(obj, level) {
      return _merge(_filtersObj, obj, level); // Return a copy of the filters
    }

    function _mergeIntoFilters(obj) {
      return _merge(_filtersObj, obj); // Return a copy of the filters
    }

    function _merge(obj1, obj2, overwriteAt) {
      // Do not modify the original object - since we are modifying this object do a deep copy
      var o1 = _.cloneDeep(obj1),
          o2 = _.clone(obj2);

      function bools(type, facet) {
        for (var bool in o2[type][facet]) {
          if (o2[type][facet].hasOwnProperty(bool) &&
              (!o1[type][facet].hasOwnProperty(bool) || overwriteAt === 'bool')) {
            o1[type][facet][bool] = o2[type][facet][bool];
          }
        }
      }

      function facets(type) {
        for (var facet in o2[type]) {
          if (o2[type].hasOwnProperty(facet) && (!o1[type].hasOwnProperty(facet) || overwriteAt === 'facet')) {
            o1[type][facet] = o2[type][facet];
          } else {
            bools(type, facet);
          }
        }
      }

      for (var type in o2) {
        if (o2.hasOwnProperty(type) && (!o1.hasOwnProperty(type) || overwriteAt === 'type')) {
          o1[type] = o2[type];
        } else {
          facets(type);
        }
      }

      return o1;
    }

    _init();

  });